### PR TITLE
[Feature] Implement Fractal Canopy and Tree Plotting with Tkinter

### LINKED ISSUE #6: Implement Recursive Fractal Drawing with Tkinter
Implement a recursive fractal drawing feature using Tkinter. The fractal should be generated based on user-defined parameters and displayed in a GUI.  

### 1. Recursive Fractal Implementation: 
- Implement the fractal generation using a recursive algorithm.
- Ensure the fractal can be drawn to a specified depth.

#### Parameters of the main (modifiable) fractal tree:
_Constant_: 
  - Start Coordinates (start_x, start_y): The starting point of the fractal—the middle of the x axis and the bottom of the y axis.
  - Angle: The initial angle of the trunk of the fractal tree—it is set constant at -90 degrees.
 
 _Modifiable:_
  - Depth: The maximum depth of recursion. Controls the complexity of the fractal.
  - Length: The initial length of the fractal branches.

  - Number of Children (num_children): The number of branches each node will have.
  - Length Scale (length_scale): The ratio of the length of child branches to the parent branch.
  - Delta Angle (delta_angle): The angle difference between consecutive branches.
  - Color (base_color and leaves_color)

### 2. Tkinter Integration (GUI):

- [x] Create a Tkinter window to display the fractal.
- [ ] Create sliders for each parameter specified as adjustable. They let the user manipulate the parameters, which should translate to the look of the fractal.
- [ ] Tie the scales to the fractal generating functions to modify the fractal as the parameters get adjusted.
- [ ]  _Optional:  Implement 2 color sliders either as buttons to choose a color or as slider to choose color hue, possibly with a visual representation of the values as a Label (perhaps).
Let the user choose the colors with Tkinter's `colorchooser` to or [use the sliders to choose the hue](https://stackoverflow.com/a/30688948/9548903) with [HSL or  HSV color map](https://en.wikipedia.org/wiki/HSL_and_HSV)  (the saturation and lightness/value would be predefined to provide best contrast)._

## Other Tasks
- [x] Implement the recursive fractal generation algorithm.
- [x] Create a Tkinter GUI with input fields for parameters and canvas with zoom feature.
- [x] Implement the drawing logic using the Canvas widget.
- [x] Write the logic to modify the fractal tree based on parameter change (by sliders) during runtime.
- [ ] Test the application to ensure it works as expected—write basic unit tests and do manual testing.

### PR DESCRIPTION
**Pull request by ola-sk**
### Overview
This pull request adds a new feature for visualizing fractal tree canopies and sine wave branches using Tkinter. The implementation introduces functions to generate fractal trees and plot them interactively on a canvas. This functionality is built on the `fractal_canopy` function and integrated with helper functions for generating gradients and drawing sine wave segments along the tree branches.

### Key Changes

- **fractal_plot_funcs.py:** Introduced the fractal_canopy function for recursively drawing fractal tree structures. The function supports multiple configuration options, including branch angles, iteration depth, color gradients, and sine wave effects.
- **helper_funcs.py:** Added helper functions such as draw_sine_wave_segment for creating sine wave effects along the branches and generate_gradient for color gradient generation between two colors.
- **fractal_test.py:** Created a testing script to visualize the fractal canopy on a Tkinter canvas. The script includes an example setup to draw a fractal tree with customizable settings like branch length, iterations, and gradient coloring.
### Context
 This feature enhances the project with an interactive visualization of fractal tree structures, which can be used for various data visualizations, simulations, or artistic displays. The fractal canopy is drawn recursively, and the tree branches can be stylized with colors and sine wave patterns for more dynamic visuals.

### Testing/Verification
- A working example is provided in fractal_test.py where a fractal tree canopy is drawn using a gradient from brown to green.
- To run the test, simply execute the script, and the fractal tree should appear on a Tkinter canvas. Adjust parameters such as n_iters, n_splits, length_ratio, and wave_amp to experiment with different tree shapes.
- Ensure that the generated gradient and sine wave effects display correctly and that the program runs without performance issues on different tree sizes.

### Screenshot
Screenshot_canopy

### Additional Notes
- The tree structure is highly customizable, and future improvements could include adding more advanced visual effects, optimizing performance, or adding more interactivity.
- Be aware that adding sine wave effects (controlled by wave_amp) can impact performance for larger trees.

### CODE REVIEW

File: fractal_test.py
Comment (ola-sk) [inline fractal_test.py:1]: I'd restrain from adding the manual test files that are for visual exploration directly to the root directory of the project. For now I would keep it on the separate branch, where everyone can checkout the changes at that specific point in time. The dev branch typically serves as a shared working branch for collaborative development. It should contain work that is relatively stable and free of temporary testing artifacts to ensure smooth team workflows, and reduce clutter for other developers.

File: fractal_plot_funcs.py [hunk 1]
```diff
@@ -1,8 +1,10 @@
 import numpy as np
 from numpy import radians as rad
 import tkinter as tk
+from helper_funcs import draw_sine_wave_segment
+from helper_funcs import generate_gradient
 
-def fractal_canopy(canvas: tk.Canvas, x: int, y: int, init_length = 200, angle_delta = 10, start_angle = -90, n_splits = 2, n_iters = 3,length_decay=2, first_iter = True) -> None:
+def fractal_canopy(canvas: tk.Canvas, x: int, y: int, init_length = 200, angle_delta = 10, start_angle = -90, n_splits = 2, n_iters = 3,length_ratio=0.75, off_angle = 0, first_iter = True) -> None:
```
Comment (ola-sk) [inline fractal_plot_funcs.py]: `length_ratio` is more readable +1

File: fractal_plot_funcs.py
Comment (ola-sk) [inline fractal_plot_funcs.py:1]: There are already performance issues when generating trees with more iterations. Once we add sliders and the parameters will be expected to change dynamically, the current delay will be unacceptable. 
We should split the tree generation part from the drawing and use optimisation techniques. We could think about precomputing the tree, caching, multithreading, batch drawing the lines with precomputed line points instead doing everything recursively. etc.

Below illustrates the benchmarked waiting time. It does not include all the waiting time, which with the parameters on the screenshot took over 6 minutes (at least, after this I stopped observing for some time, could well be 15 minutes after which the windows was still unresponsive, computations were going on and the window was hanging up. Just to let everyone know it is a real issue. With disabling the wave_amp the performance is slightly better.
image

File: fractal_funcs.py
Comment (ola-sk) [inline fractal_funcs.py:1]: Another report related to performance: with the sinevawes disabled (0) and more iterations (12) the tree gets generated after maybe 2 minutes very roughly (so it is still unacceptable) and the computations do not seem to run in infinity. Window is responsive at some point and I can drag it almost w/o a lag. 
image

File: fractal_funcs.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    angle_delta: float = 10,
+    start_angle: float = -90,
+    n_splits: int = 2,
+    n_iters: int = 3,
+    length_ratio: float = 0.75,
+    init_length: float = 200,
+    wave_amp: float = 0,
+    width: float = 1,
+    width_ratio: float = 0.75,
+    color="#000000",
```
Comment (ola-sk) [inline fractal_funcs.py:34]: we should stay consistent and add type hint to the color as well. Furthermore we should put restrictions on how many elements the gradient list has as I noticed it impacts performance when generating the tree and creates errors too. Here for example it just stretches the range if the range contains too many values: 
Screenshot 2024-11-29 193950

**Review by ola-sk — COMMENTED** [review]
We need to improve the performance of the functions generating the fractal and drawing it. We should start off by splitting the functionalities and separate the concerns of the generator and drawing functions. That will let us precompute the fractals and may enable us to draw of the precomputed branches sequentially, making the program more reactive and the window more likely to stay responsive. We could then even try to use multithreading to speed things up and make it even more responsive.

File: fractal_funcs.py (hunk 2 above)
Comment (ola-sk) [reply fractal_funcs.py:34]: Team has decided the PR can be merged in the current state. Resolving the comment.

File: fractal_funcs.py
Comment (ola-sk) [reply fractal_funcs.py:1]: Team has decided the PR can be merged in the current state. Resolving the comment.

File: fractal_test.py
Comment (ola-sk) [reply fractal_test.py:1]: Team has decided the PR can be merged in the current state. Resolving the comment.

**Comment by ola-sk** [comment]
The team has had a presentation explaining the code changes from this PR 2nd of December 2024, the following Monday the team made decision during the meeting that this PR will be merged.

**Review by Iv-Tatiana — COMMENTED** [review]
The implementation provides a solid foundation for generating and visualizing fractal tree structures with customizable features such as gradients, sine wave effects, and iteration depth. The fractal generation logic is well-structured and extensible. Performance optimization should be considered. It could be separating generation from drawing to allow for precomputation or maybe using multiple threads to perform tasks simultaneously, or anything else that could make the program run faster and more smoothly. Approved as a great step forward, with some areas for further enhancement.

**Pull request merged**